File: apps/ios/Voiid/Voiid/Networking/TokenStore.swift
//
//  TokenStore.swift
//  Voiid
//
//  Secure storage for OUR JWT (and the user id) in the iOS Keychain — never
//  UserDefaults. The chat session keys (e2e-core pickle keys) get their own
//  Keychain entries elsewhere; this is just the auth token.
//

import Foundation
import Security

/// Keychain-backed store for the auth JWT + user id. Thread-safe (Keychain is).
final class TokenStore {
    static let shared = TokenStore()

    private let jwtKey = "jwt"
    private let userIdKey = "user_id"

    private init() {}

    var jwt: String? {
        get { read(jwtKey) }
        set { newValue.map { save(jwtKey, $0) } ?? delete(jwtKey) }
    }

    var userId: String? {
        get { read(userIdKey) }
        set { newValue.map { save(userIdKey, $0) } ?? delete(userIdKey) }
    }

    var isAuthenticated: Bool { jwt != nil }

    func clear() {
        delete(jwtKey)
        delete(userIdKey)
    }

    // MARK: - Keychain primitives (shared access group so the NSE can read the JWT)

    /// Backed by the SHARED-access-group keychain store: the Notification Service
    /// Extension needs the auth JWT to call the API. Legacy app-private items are
    /// migrated into the shared group on first access (see KeychainData).
    private let kc = KeychainData(service: "com.voiid.auth")

    private func save(_ key: String, _ value: String) { kc.set(value, key) }
    private func read(_ key: String) -> String? { kc.string(key) }
    private func delete(_ key: String) { kc.delete(key) }
}

File: apps/android/app/src/main/java/com/voiid/app/net/TokenStore.kt
package com.voiid.app.net

import android.content.Context
import android.content.SharedPreferences

/**
 * Secure storage for OUR JWT + user id, mirroring iOS TokenStore (Keychain).
 * Backed by EncryptedSharedPreferences (hardware-backed master key via Keystore).
 * The e2e-core pickle keys get their own encrypted entries elsewhere.
 */
class TokenStore private constructor(private val prefs: SharedPreferences) {

    companion object {
        @Volatile private var instance: TokenStore? = null

        fun get(context: Context): TokenStore =
            instance ?: synchronized(this) {
                instance ?: TokenStore(SecurePrefs.open(context, "voiid_auth")).also { instance = it }
            }
    }

    var jwt: String?
        get() = prefs.getString("jwt", null)
        set(v) {
            prefs.edit().apply { if (v == null) remove("jwt") else putString("jwt", v) }.apply()
            // Diagnostics: which KIND of credential is held (a 1h bootstrap vs a 30-day device
            // session) and when it expires. Claims only — the token itself is never logged.
            v?.let { android.util.Log.i("VoiidAuth", "token stored: ${describe(it)}") }
        }

    private fun describe(jwt: String): String = runCatching {
        val payload = String(android.util.Base64.decode(jwt.split(".")[1],
            android.util.Base64.URL_SAFE or android.util.Base64.NO_PADDING or android.util.Base64.NO_WRAP))
        val o = org.json.JSONObject(payload)
        "scope=${o.optString("scope", "legacy")} exp=${java.util.Date(o.optLong("exp") * 1000)}"
    }.getOrDefault("unparseable")

    var userId: String?
        get() = prefs.getString("user_id", null)
        set(v) { prefs.edit().apply { if (v == null) remove("user_id") else putString("user_id", v) }.apply() }

    val isAuthenticated: Boolean get() = jwt != null

    fun clear() {
        android.util.Log.w("VoiidAuth", "token cleared", Throwable("who cleared it"))
        prefs.edit().remove("jwt").remove("user_id").apply()
    }
}
